File: apps/web/src/api/splunk_live.ts
export interface SplunkLiveStatus {
  configured: boolean
  connected: boolean
  server_info: {
    version: string
    product_name: string
    server_name: string
    os: string
  } | null
  error: string | null
  warning: string
}

export interface SplunkLiveQuery {
  id: number
  case_id: number | null
  spl_query: string
  template_id: string | null
  earliest_time: string | null
  latest_time: string | null
  result_count: number
  result_sample: string | null
  status: string
  error_message: string | null
  splunk_host: string | null
  executed_at: string
  warning: string
}

export interface SplunkLiveSearchRequest {
  spl_query: string
  case_id?: number
  earliest_time?: string
  latest_time?: string
  max_results?: number
}

export interface SplunkLiveTemplateRunRequest {
  case_id?: number
  earliest_time?: string
  latest_time?: string
  max_results?: number
}

export async function getSplunkLiveStatus(): Promise<SplunkLiveStatus> {
  const res = await fetch('/api/splunk/live/status')
  if (!res.ok) throw new Error(`Status check failed: ${res.status}`)
  return res.json()
}

export async function runSplunkLiveSearch(
  req: SplunkLiveSearchRequest,
): Promise<SplunkLiveQuery> {
  const res = await fetch('/api/splunk/live/search', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(req),
  })
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }))
    throw new Error(err.detail ?? 'Live search failed')
  }
  return res.json()
}

export async function runSplunkLiveTemplate(
  templateId: string,
  req: SplunkLiveTemplateRunRequest = {},
): Promise<SplunkLiveQuery> {
  const res = await fetch(`/api/splunk/live/templates/${templateId}/run`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(req),
  })
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }))
    throw new Error(err.detail ?? 'Template run failed')
  }
  return res.json()
}

export async function getSplunkLiveQueries(caseId?: number): Promise<SplunkLiveQuery[]> {
  const url = caseId
    ? `/api/splunk/live/queries?case_id=${caseId}`
    : '/api/splunk/live/queries'
  const res = await fetch(url)
  if (!res.ok) throw new Error(`Failed to fetch query history: ${res.status}`)
  return res.json()
}

File: services/api/routers/splunk_live.py
"""
Live Splunk Connector endpoints.

GET  /splunk/live/status                       — connection status and server info
POST /splunk/live/search                       — run a free-form SPL query
POST /splunk/live/templates/{template_id}/run  — run a pre-approved SPL template
GET  /splunk/live/queries                      — list stored query history

Credentials are read from SPLUNK_URL and SPLUNK_TOKEN environment variables.
No credentials are stored in the database — only hostname and query metadata.

WARNING: Do not use production credentials in development or demo environments.
"""

import json
import os
import sys
from pathlib import Path
from typing import List, Optional
from urllib.parse import urlparse

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

_REPO_ROOT = Path(__file__).resolve().parent.parent.parent.parent
if str(_REPO_ROOT) not in sys.path:
    sys.path.insert(0, str(_REPO_ROOT))

from integrations.splunk.connector import (  # noqa: E402
    is_configured,
    test_connection,
    run_search,
    SAFETY_WARNING,
)
from integrations.splunk.query_templates import QUERY_TEMPLATES  # noqa: E402

from database import get_db
from models.splunk_live_query import SplunkLiveQuery
from schemas.splunk_live_schema import (
    SplunkStatusResponse,
    SplunkLiveSearchRequest,
    SplunkLiveTemplateRunRequest,
    SplunkLiveQueryResponse,
)

router = APIRouter(tags=["splunk-live"])


@router.get("/splunk/live/status", response_model=SplunkStatusResponse)
def get_splunk_live_status():
    """Test the live Splunk connection and return connection status."""
    result = test_connection()
    return SplunkStatusResponse(
        configured=result["configured"],
        connected=result["connected"],
        server_info=result.get("server_info"),
        error=result.get("error"),
        warning=result["warning"],
    )


@router.post("/splunk/live/search", response_model=SplunkLiveQueryResponse)
def run_splunk_live_search(
    body: SplunkLiveSearchRequest,
    db: Session = Depends(get_db),
):
    """
    Run a free-form SPL query against a live Splunk instance.
    Stores query metadata and a sample of up to 20 results — not the full result set.
    """
    if not is_configured():
        raise HTTPException(
            status_code=503,
            detail=(
                "Live Splunk connector is not configured. "
                "Set SPLUNK_URL and SPLUNK_TOKEN environment variables."
            ),
        )

    splunk_host = _extract_host(os.environ.get("SPLUNK_URL", ""))

    row = SplunkLiveQuery(
        case_id=body.case_id,
        spl_query=body.spl_query,
        template_id=None,
        earliest_time=body.earliest_time,
        latest_time=body.latest_time,
        status="pending",
        splunk_host=splunk_host,
    )
    db.add(row)
    db.commit()
    db.refresh(row)

    result = run_search(body.spl_query, body.earliest_time, body.latest_time, body.max_results)
    _apply_result(db, row, result)

    return _build_response(row)


@router.post(
    "/splunk/live/templates/{template_id}/run",
    response_model=SplunkLiveQueryResponse,
)
def run_splunk_live_template(
    template_id: str,
    body: SplunkLiveTemplateRunRequest,
    db: Session = Depends(get_db),
):
    """
    Run a pre-approved SPL template against a live Splunk instance.
    Only templates from the built-in library can be run this way.
    """
    if not is_configured():
        raise HTTPException(
            status_code=503,
            detail=(
                "Live Splunk connector is not configured. "
                "Set SPLUNK_URL and SPLUNK_TOKEN environment variables."
            ),
        )

    template = next((t for t in QUERY_TEMPLATES if t["id"] == template_id), None)
    if not template:
        raise HTTPException(status_code=404, detail=f"Template '{template_id}' not found.")

    splunk_host = _extract_host(os.environ.get("SPLUNK_URL", ""))

    row = SplunkLiveQuery(
        case_id=body.case_id,
        spl_query=template["spl_query"],
        template_id=template_id,
        earliest_time=body.earliest_time,
        latest_time=body.latest_time,
        status="pending",
        splunk_host=splunk_host,
    )
    db.add(row)
    db.commit()
    db.refresh(row)

    result = run_search(
        template["spl_query"],
        body.earliest_time,
        body.latest_time,
        body.max_results,
    )
    _apply_result(db, row, result)

    return _build_response(row)


@router.get("/splunk/live/queries", response_model=List[SplunkLiveQueryResponse])
def list_splunk_live_queries(
    case_id: Optional[int] = None,
    db: Session = Depends(get_db),
):
    """List stored live query history, newest first. Optionally filtered by case."""
    q = db.query(SplunkLiveQuery).order_by(SplunkLiveQuery.executed_at.desc())
    if case_id is not None:
        q = q.filter(SplunkLiveQuery.case_id == case_id)
    return [_build_response(r) for r in q.limit(100).all()]


# ── helpers ────────────────────────────────────────────────────────────────────

def _extract_host(url: str) -> str:
    """Extract hostname from URL for safe logging (no token, no credentials)."""
    try:
        parsed = urlparse(url)
        return parsed.hostname or url
    except Exception:
        return ""


def _apply_result(db: Session, row: SplunkLiveQuery, result: dict) -> None:
    """Write search result back to the row and commit."""
    sample = json.dumps(result["results"][:20], default=str) if result.get("results") else None
    row.result_count = result.get("result_count", 0)
    row.result_sample = sample
    row.status = "error" if result.get("error") else "completed"
    row.error_message = result.get("error")
    db.commit()
    db.refresh(row)


def _build_response(row: SplunkLiveQuery) -> SplunkLiveQueryResponse:
    return SplunkLiveQueryResponse(
        id=row.id,
        case_id=row.case_id,
        spl_query=row.spl_query,
        template_id=row.template_id,
        earliest_time=row.earliest_time,
        latest_time=row.latest_time,
        result_count=row.result_count or 0,
        result_sample=row.result_sample,
        status=row.status,
        error_message=row.error_message,
        splunk_host=row.splunk_host,
        executed_at=row.executed_at.isoformat() if row.executed_at else "",
        warning=SAFETY_WARNING,
    )
